engine/apps/slack/scenarios/distribute_alerts.py

```diff
@@ -191,75 +263,6 @@ def process_signal(self, alert: Alert) -> None:
-            logger.info(
-                f"Failed to post message to Slack for alert_group {alert_group.pk} because slack_channel is None"
-            )
-
-            alert_group.reason_to_skip_escalation = AlertGroup.CHANNEL_NOT_SPECIFIED
-            alert_group.save(update_fields=["reason_to_skip_escalation"])
-
-            return
-
-        try:
-            result = self._slack_client.chat_postMessage(
-                channel=slack_channel.slack_id,
-                attachments=attachments,
-                blocks=blocks,
-            )
-
-            # TODO: once _channel_id has been fully migrated to channel, remove _channel_id
-            # see https://raintank-corp.slack.com/archives/C06K1MQ07GS/p1732555465144099
-            alert_group.slack_messages.create(
-                slack_id=result["ts"],
-                organization=alert_group.channel.organization,
-                _channel_id=slack_channel.slack_id,
-                channel=slack_channel,
-            )
-
-            alert.delivered = True
-        except SlackAPITokenError:
-            alert_group.reason_to_skip_escalation = AlertGroup.ACCOUNT_INACTIVE
-            alert_group.save(update_fields=["reason_to_skip_escalation"])
-            logger.info("Not delivering alert due to account_inactive.")
-        except SlackAPIInvalidAuthError:
-            alert_group.reason_to_skip_escalation = AlertGroup.INVALID_AUTH
-            alert_group.save(update_fields=["reason_to_skip_escalation"])
-            logger.info("Not delivering alert due to invalid_auth.")
-        except SlackAPIChannelArchivedError:
-            alert_group.reason_to_skip_escalation = AlertGroup.CHANNEL_ARCHIVED
-            alert_group.save(update_fields=["reason_to_skip_escalation"])
-            logger.info("Not delivering alert due to channel is archived.")
-        except SlackAPIRatelimitError as e:
-            # don't rate limit maintenance alert
-            if not alert_group.channel.is_maintenace_integration:
-                alert_group.reason_to_skip_escalation = AlertGroup.RATE_LIMITED
-                alert_group.save(update_fields=["reason_to_skip_escalation"])
-                alert_group.channel.start_send_rate_limit_message_task(e.retry_after)
-                logger.info("Not delivering alert due to slack rate limit.")
-            else:
-                raise e
-        except SlackAPIChannelNotFoundError:
-            alert_group.reason_to_skip_escalation = AlertGroup.CHANNEL_ARCHIVED
-            alert_group.save(update_fields=["reason_to_skip_escalation"])
-            logger.info("Not delivering alert due to channel is archived.")
-        except SlackAPIRestrictedActionError:
-            alert_group.reason_to_skip_escalation = AlertGroup.RESTRICTED_ACTION
-            alert_group.save(update_fields=["reason_to_skip_escalation"])
-            logger.info("Not delivering alert due to workspace restricted action.")
-        finally:
-            alert.save()
```

Comment:
(I also just moved all of the logic that was in `_post_alert_group_to_slack` inside of `process_signal`, as it was more straightforward to do all of the exception handling there directly 🙂

Reply:
The main part makes a lot of sense to me 👍 
I'd propose to recheck the part with updating `slack_message_sent` field. We reset it to False _only_ if we reraise the exception 🤔 I think the idea here is to prevent attempts to post a message to slack in case when we are sure it's not possible (e.g. slack token error), because we call this step with every new alert in the alert group. So with every next alert in the alert group `num_updated_rows` should return `0`, and we should skip "post the first message" part. Does it make sense?

Reply:
good point, thanks for catching this 👍 updated in https://github.com/grafana/oncall/pull/5314/commits/62b5a85523baa559f1e349d85562cc94875481d9 (and updated the tests to assert that this value is set to the proper value, for all of the various `Exception`s that we handle here)